types/src/params.rs

```diff
@@ -408,6 +416,275 @@ impl<'a> Id<'a> {
+			unsafe { String::from_utf8_unchecked(self.bytes) }
+		}
+	}
+}
+
+/// Parameter builder that serializes named value parameters to a JSON compatible string.
+/// This is the equivalent of a JSON Map object `{ key: value }`.
+///
+/// # Examples
+///
+/// ```rust
+///
+/// use jsonrpsee_types::NamedParamsBuilder;
+///
+/// let mut builder = NamedParamsBuilder::new();
+/// builder.insert("param1", 1);
+/// builder.insert("param2", "abc");
+/// let params = builder.build();
+///
+/// // Use RPC parameters...
+/// ```
+#[derive(Debug)]
+pub struct NamedParamsBuilder(params_builder::ParamsBuilder);
+
+impl NamedParamsBuilder {
+	/// Construct a new [`NamedParamsBuilder`].
+	pub fn new() -> Self {
+		Self::default()
+	}
+
+	/// Insert a named value (key, value) pair into the builder.
+	/// The _name_ and _value_ are delimited by the `:` token.
+	pub fn insert<P: Serialize>(&mut self, name: &str, value: P) -> Result<(), serde_json::Error> {
+		self.0.insert_named(name, value)
+	}
+
+	/// Finish the building process and return a JSON compatible string.
+	pub fn build(self) -> NamedParams {
+		NamedParams(self.0.build())
+	}
+}
+
+impl Default for NamedParamsBuilder {
+	fn default() -> Self {
+		Self(params_builder::ParamsBuilder::named())
+	}
+}
+
+/// Named RPC parameters stored as a JSON Map object `{ key: value }`.
+#[derive(Clone, Debug)]
+pub struct NamedParams(String);
+
+impl ToRpcParams for NamedParams {
+	fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
+		RawValue::from_string(self.0).map(Some)
+	}
+}
+
+/// Parameter builder that serializes plain value parameters to a JSON compatible string.
+/// This is the equivalent of a JSON Array object `[ value0, value1, .., valueN ]`.
+///
+/// # Examples
+///
+/// ```rust
+///
+/// use jsonrpsee_types::UnnamedParamsBuilder;
+///
+/// let mut builder = UnnamedParamsBuilder::new();
+/// builder.insert("param1");
+/// builder.insert(1);
+/// let params = builder.build();
+///
+/// // Use RPC parameters...
+/// ```
+#[derive(Debug)]
+pub struct UnnamedParamsBuilder(params_builder::ParamsBuilder);
+
+impl UnnamedParamsBuilder {
+	/// Construct a new [`UnnamedParamsBuilder`].
+	pub fn new() -> Self {
+		Self::default()
+	}
+
+	/// Insert a plain value into the builder.
+	pub fn insert<P: Serialize>(&mut self, value: P) -> Result<(), serde_json::Error> {
+		self.0.insert(value)
+	}
+
+	/// Finish the building process and return a JSON compatible string.
+	pub fn build(self) -> UnnamedParams {
+		UnnamedParams(self.0.build())
+	}
+}
+
+impl Default for UnnamedParamsBuilder {
+	fn default() -> Self {
+		Self(params_builder::ParamsBuilder::positional())
+	}
+}
+
+/// Unnamed RPC parameters stored as a JSON Array object `[ value0, value1, .., valueN ]`.
+#[derive(Clone, Debug)]
+pub struct UnnamedParams(String);
+
+impl ToRpcParams for UnnamedParams {
+	fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
```

Comment:
Nit:

Just thinking about other people implementing this; who knows what errors they might want to express in their implementations.

Is it worth doing something like:

```rust
trait ToRpcParams {
    fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, jsonrpsee::Error>;
}
```

So that people have more control over what sorts of errors they spit out?

Reply:
I do like the idea of exposing a more generic error here. The only inconvenient I can think of is that the general [jsonrpsee::Error](https://github.com/paritytech/jsonrpsee/blob/49f260eb3ce41ad90704f783039add98661cde5f/core/src/error.rs#L59) is defined in the `core` create. 

Another good point for returning the `jsonrpsee::Error` is that's the same error we expect to return from the [client's trait](https://github.com/paritytech/jsonrpsee/blob/49f260eb3ce41ad90704f783039add98661cde5f/core/src/client/mod.rs#L58).

To achieve that, I believe we should move the `core/src/error` to `types/src/error`. Currently, we are limited by the fact the `core` crate depends on the `types` crate, and we'd want the `types` crate to also depend on `core::Error`.

I suggest we should move the files if possible. @jsdw @niklasad1 What do you think about it?